Migrate ServiceWorkerRegistration data to origin directory

szewai opened the pull request:
#### 1e35cedbb5b8c4d52b35faccb1b0b2ac7b619878
<pre>
Migrate ServiceWorkerRegistration data to origin directory
<a href="https://bugs.webkit.org/show_bug.cgi?id=254771">https://bugs.webkit.org/show_bug.cgi?id=254771</a>
rdar://106385894

Reviewed by Youenn Fablet.

ServiceWorkerRegistration data is currently placed in Caches directory by default (see
WebsiteDataStore::defaultServiceWorkerRegistrationDirectory), which means ServiceWorker registrations could be deleted
by system. To avoid that, we should move it to a non-Caches directory.

To make ServiceWorkerRegistration aligned with other storage types like CacheStorage and IndexedDB, this patch adds
functions to migrate ServiceWorkerRegistration data to origin directory. Origin directory is managed by
OriginStorageManager, so for accessing ServiceWorkerRegistration data, SWServer will need to talk to
OriginStorageManager, or NetworkStorageManager (the owner of OriginStorageManager). This is implemented by replacing
RegistrationStore with SWRegistrationStore, and having SWRegistrationStore (implemented in WebSWRegistrationStore)
pointing to NetworkStorageManager.

A new class ServiceWorkerStorageManager is added to manage ServiceWorkerRegistration data of one &quot;partition&quot;. Before
migration, ServiceWorkerRegistration has one database for all origins (i.e. there is only one &quot;partition&quot; for all
origins), so NetworkStorageManager owns a ServiceWorkerStorageManager. After migration, each origin has a database (i.e.
there is one &quot;partition&quot; per origin), so each OriginStorageManager owns a ServiceWorkerStorageManager and thus a
SWRegistrationDatabase. SWRegistrationDatabase is the single-thread version of the existing
RegistrationDatabase, because ServiceWorkerStorageManager only runs on WorkQueue of NetworkStorageManager.

New test: WKWebsiteDataStore.MigrateServiceWorkerRegistrationToGeneralStorageDirectory

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/WebCorePersistentCoders.cpp:
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/ServiceWorkerTypes.h:
(WebCore::ServiceWorkerScripts::isolatedCopy const):
* Source/WebCore/workers/service/server/RegistrationDatabase.cpp:
(WebCore::recordsTableSchema): Deleted.
(WebCore::recordsTableSchemaAlternate): Deleted.
(WebCore::databaseFilenameFromVersion): Deleted.
(WebCore::databaseFilename): Deleted.
(WebCore::serviceWorkerRegistrationDatabaseFilename): Deleted.
(WebCore::cleanOldDatabases): Deleted.
(WebCore::stripScriptSources): Deleted.
(WebCore::populateScriptSourcesFromDisk): Deleted.
(WebCore::registrationDatabaseWorkQueue): Deleted.
(WebCore::RegistrationDatabase::RegistrationDatabase): Deleted.
(WebCore::RegistrationDatabase::~RegistrationDatabase): Deleted.
(WebCore::RegistrationDatabase::postTaskToWorkQueue): Deleted.
(WebCore::RegistrationDatabase::openSQLiteDatabase): Deleted.
(WebCore::RegistrationDatabase::scriptStorageDirectory const): Deleted.
(WebCore::RegistrationDatabase::scriptStorage): Deleted.
(WebCore::RegistrationDatabase::importRecordsIfNecessary): Deleted.
(WebCore::RegistrationDatabase::ensureValidRecordsTable): Deleted.
(WebCore::updateViaCacheToString): Deleted.
(WebCore::stringToUpdateViaCache): Deleted.
(WebCore::workerTypeToString): Deleted.
(WebCore::stringToWorkerType): Deleted.
(WebCore::RegistrationDatabase::pushChanges): Deleted.
(WebCore::RegistrationDatabase::schedulePushChanges): Deleted.
(WebCore::RegistrationDatabase::close): Deleted.
(WebCore::RegistrationDatabase::clearAll): Deleted.
(WebCore::RegistrationDatabase::doPushChanges): Deleted.
(WebCore::RegistrationDatabase::doPushChangesWithOpenDatabase): Deleted.
(WebCore::RegistrationDatabase::importRecords): Deleted.
(WebCore::RegistrationDatabase::addRegistrationToStore): Deleted.
(WebCore::RegistrationDatabase::databaseFailedToOpen): Deleted.
(WebCore::RegistrationDatabase::databaseOpenedAndRecordsImported): Deleted.
* Source/WebCore/workers/service/server/RegistrationDatabase.h:
(WebCore::RegistrationDatabase::create): Deleted.
(WebCore::RegistrationDatabase::databaseDirectoryIsolatedCopy const): Deleted.
(): Deleted.
* Source/WebCore/workers/service/server/RegistrationStore.cpp:
(WebCore::RegistrationStore::RegistrationStore): Deleted.
(WebCore::RegistrationStore::~RegistrationStore): Deleted.
(WebCore::RegistrationStore::scheduleDatabasePushIfNecessary): Deleted.
(WebCore::RegistrationStore::pushChangesToDatabase): Deleted.
(WebCore::RegistrationStore::clearAll): Deleted.
(WebCore::RegistrationStore::flushChanges): Deleted.
(WebCore::RegistrationStore::startSuspension): Deleted.
(WebCore::RegistrationStore::closeDatabase): Deleted.
(WebCore::RegistrationStore::endSuspension): Deleted.
(WebCore::RegistrationStore::updateRegistration): Deleted.
(WebCore::RegistrationStore::removeRegistration): Deleted.
(WebCore::RegistrationStore::addRegistrationFromDatabase): Deleted.
(WebCore::RegistrationStore::didSaveWorkerScriptsToDisk): Deleted.
(WebCore::RegistrationStore::databaseFailedToOpen): Deleted.
(WebCore::RegistrationStore::databaseOpenedAndRecordsImported): Deleted.
* Source/WebCore/workers/service/server/RegistrationStore.h:
(WebCore::RegistrationStore::server): Deleted.
(WebCore::RegistrationStore::pushChangesToDatabase): Deleted.
(): Deleted.
* Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp: Added.
(WebCore::databaseFilePath):
(WebCore::scriptDirectoryPath):
(WebCore::convertUpdateViaCacheToString):
(WebCore::convertStringToUpdateViaCache):
(WebCore::convertWorkerTypeToString):
(WebCore::convertStringToWorkerType):
(WebCore::currentRecordsTableSchema):
(WebCore::currentRecordsTableSchemaAlternate):
(WebCore::stripScriptSources):
(WebCore::populateScriptSourcesFromDisk):
(WebCore::SWRegistrationDatabase::statementString const):
(WebCore::SWRegistrationDatabase::cachedStatement):
(WebCore::SWRegistrationDatabase::SWRegistrationDatabase):
(WebCore::SWRegistrationDatabase::~SWRegistrationDatabase):
(WebCore::SWRegistrationDatabase::scriptStorage):
(WebCore::SWRegistrationDatabase::prepareDatabase):
(WebCore::SWRegistrationDatabase::ensureValidRecordsTable):
(WebCore::SWRegistrationDatabase::importRegistrations):
(WebCore::SWRegistrationDatabase::updateRegistrations):
* Source/WebCore/workers/service/server/SWRegistrationDatabase.h: Added.
* Source/WebCore/workers/service/server/SWRegistrationStore.h: Added.
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::storeRegistrationsOnDisk):
(WebCore::SWServer::SWServer):
(WebCore::SWServer::Connection::storeRegistrationsOnDisk):
(WebCore::SWServer::startSuspension): Deleted.
(WebCore::SWServer::endSuspension): Deleted.
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerDelegate.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::prepareToSuspend):
(WebKit::NetworkProcess::processDidResume):
(WebKit::NetworkProcess::storeServiceWorkerRegistrations):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
(WebKit::NetworkSession::createUniqueRegistrationStore):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWRegistrationStore.cpp: Added.
(WebKit::WebSWRegistrationStore::WebSWRegistrationStore):
(WebKit::WebSWRegistrationStore::clearAll):
(WebKit::WebSWRegistrationStore::flushChanges):
(WebKit::WebSWRegistrationStore::closeFiles):
(WebKit::WebSWRegistrationStore::importRegistrations):
(WebKit::WebSWRegistrationStore::updateRegistration):
(WebKit::WebSWRegistrationStore::removeRegistration):
(WebKit::WebSWRegistrationStore::scheduleUpdateIfNecessary):
(WebKit::WebSWRegistrationStore::updateToStorage):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWRegistrationStore.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::allManagedTypes):
(WebKit::NetworkStorageManager::closeServiceWorkerRegistrationFiles):
(WebKit::NetworkStorageManager::clearServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::importServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::updateServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::migrateServiceWorkerRegistrationsToOrigins):
(WebKit::NetworkStorageManager::updateServiceWorkerRegistrationsByOrigin):
(WebKit::NetworkStorageManager::shouldManageServiceWorkerRegistrationsByOrigin):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::toStorageType const):
(WebKit::OriginStorageManager::StorageBucket::toStorageIdentifier const):
(WebKit::OriginStorageManager::StorageBucket::serviceWorkerStorageManager):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::serviceWorkerStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/NetworkProcess/storage/ServiceWorkerStorageManager.cpp: Added.
(WebKit::ServiceWorkerStorageManager::ServiceWorkerStorageManager):
(WebKit::ServiceWorkerStorageManager::ensureDatabase):
(WebKit::ServiceWorkerStorageManager::closeFiles):
(WebKit::ServiceWorkerStorageManager::clearAllRegistrations):
(WebKit::ServiceWorkerStorageManager::importRegistrations):
(WebKit::ServiceWorkerStorageManager::updateRegistrations):
* Source/WebKit/NetworkProcess/storage/ServiceWorkerStorageManager.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _storeServiceWorkerRegistrations:]):
(-[WKWebsiteDataStore _hasRegisteredServiceWorker]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::storeServiceWorkerRegistrations):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Badging.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(log):
(install):
(addEventListener):
(then):
(catch):

Canonical link: <a href="https://commits.webkit.org/262631@main">https://commits.webkit.org/262631@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/908e3f23c1f8069ddcf21ac4ee1afe547c737de4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2067 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2993 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2130 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2042 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2176 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2138 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1889 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2086 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1853 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1852 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2843 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1838 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1744 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1903 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1876 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3015 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1899 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1696 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1839 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/528 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2005 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cf585e486f59f00170cefd0f895d49bdda31ded3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/069c7ad3ff6802a5132be838ee8d5370a04de307. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9a83f21ac229469f9c68154bf82080718e53f51e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/11a5119a2f22a9eb3e67984ffc405edd2b50175d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/11a5119a2f22a9eb3e67984ffc405edd2b50175d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b3806c2b02ed1b811f0936b986eff7ae08835a62. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/315cb0988abe4d1fa0e6692ed47a5b5baacb20ef. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

youennf reviewed (commented):
Overall direction seems good to me.
It would be good to have some info on how slower for the first load this will make things in case we have like 20 files to load instead of just 1.
In the future, we might want to compute the list of all origins that have a service worker but load them per origin on the fly. This will hopefully speed up the launch of the service worker, with almost no penalty for page load given we have preloaders in case of service workers.

szewai commented on Source/WebCore/workers/service/server/SWRegistrationStore.h:
```diff
@@ -0,0 +1,49 @@
+class SWServer;
+class SWServerRegistration;
+
+class SWRegistrationStore {
```
Yes `SWRegistrationStore` is similar to `RegistrationStore` but is a virtual class (implementation is in `Web SWRegistrationStore` as we need access to `NetworkStorageManager`). I will remove `RegistrationStore.h/cpp`.

szewai commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -40084,6 +40092,8 @@
 				C11A9ECE21403A5C00CFB20A /* SwitchingGPUClient.h in Headers */,
 				4151E5BB1FBA4C7A00E47E2D /* SWOriginStore.h in Headers */,
+				3A7D62DA29D49EC900D57DAC /* SWRegistrationDatabase.h in Headers */,
+				3AC4C04429D345D800402716 /* SWRegistrationStore.h in Headers */,
```
Will remove.

szewai commented on Source/WebKit/NetworkProcess/storage/ServiceWorkerStorageManager.cpp:
```diff
@@ -0,0 +1,77 @@
+WebCore::SWRegistrationDatabase* ServiceWorkerStorageManager::ensureDatabase()
+{
+    if (!m_database && !m_path.isEmpty())
+        m_database = makeUnique<WebCore::SWRegistrationDatabase>(m_path);
```
`SWRegistrationDatabase` is a single-thread version of  `RegistrationDatabase`, which has a `WorkQueue`. I will drop `RegistrationDatabase` as suggested.

It's not quite a `OriginPartitionedRegistrationDatabase` as it is used before and after migration (like `ServiceWorkerRegistrationManager`); before migration there's only one database per session.

szewai commented on Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp:
```diff
@@ -0,0 +1,460 @@
+        if (statement->bindText(1, data.registration.key.toDatabaseKey()) != SQLITE_OK
+            || statement->bindText(2, data.registration.scopeURL.protocolHostAndPort()) != SQLITE_OK
+            || statement->bindText(3, data.registration.scopeURL.path().toString()) != SQLITE_OK
+            || statement->bindText(4, data.registration.key.topOrigin().databaseIdentifier()) != SQLITE_OK
```
That's true, but `SWRegistrationDatabase` is used for both before (one database per session) and after migration (one database per origin), so I didn't change database schema.

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -171,10 +176,17 @@ NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, Markable<
+        if (shouldManageServiceWorkerRegistrationsByOrigin())
+            migrateServiceWorkerRegistrationsToOrigins();
+        else
+            m_sharedServiceWorkerStorageManager = makeUnique<ServiceWorkerStorageManager>(m_customServiceWorkerStoragePath);
```
`RegistrationDatabase` is cross-thread and has a few issues like `RegistrationDatabase::doPushChangesWithOpenDatabase` invokes 'didSaveWorkerScriptsToDisk' before transaction commits, so I just created a simplified single-thread version -- `SWRegistrationDatabase` that can be used for before and after migration.

We might create  `SharedSWRegistrationDatabase` and `OriginSWRegistrationDatabase` and use them in different cases. Do you this is better?

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1495,6 +1513,159 @@ void NetworkStorageManager::notifyBackgroundFetchChange(const String& identifier
+        std::optional<Vector<WebCore::ServiceWorkerContextData>> crossThreadResult;
+        if (result)
+            crossThreadResult = crossThreadCopy(WTFMove(*result));
+        RunLoop::main().dispatch([result = WTFMove(crossThreadResult), completionHandler = WTFMove(completionHandler)]() mutable {
```
There is `NetworkStorageManager::close()` which should dispatch the last task to work queue

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/93395a806871b6b6df7f27a8dcb8fcadd60e6c1d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3a04a2bb780d371fa3852a1126914f7772524beb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3a04a2bb780d371fa3852a1126914f7772524beb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/de0e87e14d2013d1dc624cb6a3603a90a96014ea. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/de0e87e14d2013d1dc624cb6a3603a90a96014ea. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/15a44a996cc11c62ef9396448b4701aec45bbd34. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8bfa05307a2982dd2d57eccde486ae7e7771105b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ea2ae392188f2a5c0fa472d6ae812f9541df6096. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f31a0719600055848dbe558469797dddc41bc0b2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/29a659ff479e4ebf67fce90b68f05b951c182f14. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/33eb5a91135afc7380aa661725663dc04e3ba7c5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1495,6 +1515,152 @@ void NetworkStorageManager::notifyBackgroundFetchChange(const String& identifier
+        else
+            result = updateServiceWorkerRegistrationsByOrigin(WTFMove(registrationsToUpdate), WTFMove(registrationsToDelete));
+
+        RunLoop::main().dispatch([result = crossThreadCopy(WTFMove(result)), completionHandler = WTFMove(completionHandler)]() mutable {
```
Need to get back protectedThis to main thread.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1495,6 +1515,152 @@ void NetworkStorageManager::notifyBackgroundFetchChange(const String& identifier
+                result = registrations;
+        }
+
+        RunLoop::main().dispatch([result = crossThreadCopy(WTFMove(result)), completionHandler = WTFMove(completionHandler)]() mutable {
```
Need to get back protectedThis to main thread.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1495,6 +1515,152 @@ void NetworkStorageManager::notifyBackgroundFetchChange(const String& identifier
+            }
+        }
+
+        RunLoop::main().dispatch(WTFMove(completionHandler));
```
Need to get back protectedThis to main thread.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -1495,6 +1515,152 @@ void NetworkStorageManager::notifyBackgroundFetchChange(const String& identifier
+                manager->serviceWorkerStorageManager().closeFiles();
+        }
+
+        RunLoop::main().dispatch(WTFMove(completionHandler));
```
Need to get back protectedThis to main thread.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -57,6 +55,17 @@
+#endif
+
+#if !PLATFORM(COCOA)
+#include <wtf/text/StringBuilder.h>
```
Needed?

szewai replied:
Nope.. will remove

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -304,20 +313,7 @@ void SWServer::clearAll(CompletionHandler<void()>&& completionHandler)
-    if (m_registrationStore)
-        m_registrationStore->endSuspension();
-}
+// TODO: suspend registration database via NetworkStorageManager
```
What adoption do we need here?
Please file a bug if we need to do anything.

szewai replied:
Oh I forgot to drop this. The storage queue of `NetworkStorageManager`, which `ServiceWorkerStorageManager` runs on will stop processing task when `NetworkStorageManager::suspend`, so there is special handling needed for `ServiceWorkerStorageManager` (because `SWRegistrationDatabase` does not create transaction that outlives WorkQueue task).

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/908e3f23c1f8069ddcf21ac4ee1afe547c737de4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12204

szewai committed:
Migrate ServiceWorkerRegistration data to origin directory
https://bugs.webkit.org/show_bug.cgi?id=254771
rdar://106385894

Reviewed by Youenn Fablet.

ServiceWorkerRegistration data is currently placed in Caches directory by default (see
WebsiteDataStore::defaultServiceWorkerRegistrationDirectory), which means ServiceWorker registrations could be deleted
by system. To avoid that, we should move it to a non-Caches directory.

To make ServiceWorkerRegistration aligned with other storage types like CacheStorage and IndexedDB, this patch adds
functions to migrate ServiceWorkerRegistration data to origin directory. Origin directory is managed by
OriginStorageManager, so for accessing ServiceWorkerRegistration data, SWServer will need to talk to
OriginStorageManager, or NetworkStorageManager (the owner of OriginStorageManager). This is implemented by replacing
RegistrationStore with SWRegistrationStore, and having SWRegistrationStore (implemented in WebSWRegistrationStore)
pointing to NetworkStorageManager.

A new class ServiceWorkerStorageManager is added to manage ServiceWorkerRegistration data of one "partition". Before
migration, ServiceWorkerRegistration has one database for all origins (i.e. there is only one "partition" for all
origins), so NetworkStorageManager owns a ServiceWorkerStorageManager. After migration, each origin has a database (i.e.
there is one "partition" per origin), so each OriginStorageManager owns a ServiceWorkerStorageManager and thus a
SWRegistrationDatabase. SWRegistrationDatabase is the single-thread version of the existing
RegistrationDatabase, because ServiceWorkerStorageManager only runs on WorkQueue of NetworkStorageManager.

New test: WKWebsiteDataStore.MigrateServiceWorkerRegistrationToGeneralStorageDirectory

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/WebCorePersistentCoders.cpp:
* Source/WebCore/workers/service/ServiceWorkerRegistrationKey.h:
* Source/WebCore/workers/service/ServiceWorkerTypes.h:
(WebCore::ServiceWorkerScripts::isolatedCopy const):
* Source/WebCore/workers/service/server/RegistrationDatabase.cpp:
(WebCore::recordsTableSchema): Deleted.
(WebCore::recordsTableSchemaAlternate): Deleted.
(WebCore::databaseFilenameFromVersion): Deleted.
(WebCore::databaseFilename): Deleted.
(WebCore::serviceWorkerRegistrationDatabaseFilename): Deleted.
(WebCore::cleanOldDatabases): Deleted.
(WebCore::stripScriptSources): Deleted.
(WebCore::populateScriptSourcesFromDisk): Deleted.
(WebCore::registrationDatabaseWorkQueue): Deleted.
(WebCore::RegistrationDatabase::RegistrationDatabase): Deleted.
(WebCore::RegistrationDatabase::~RegistrationDatabase): Deleted.
(WebCore::RegistrationDatabase::postTaskToWorkQueue): Deleted.
(WebCore::RegistrationDatabase::openSQLiteDatabase): Deleted.
(WebCore::RegistrationDatabase::scriptStorageDirectory const): Deleted.
(WebCore::RegistrationDatabase::scriptStorage): Deleted.
(WebCore::RegistrationDatabase::importRecordsIfNecessary): Deleted.
(WebCore::RegistrationDatabase::ensureValidRecordsTable): Deleted.
(WebCore::updateViaCacheToString): Deleted.
(WebCore::stringToUpdateViaCache): Deleted.
(WebCore::workerTypeToString): Deleted.
(WebCore::stringToWorkerType): Deleted.
(WebCore::RegistrationDatabase::pushChanges): Deleted.
(WebCore::RegistrationDatabase::schedulePushChanges): Deleted.
(WebCore::RegistrationDatabase::close): Deleted.
(WebCore::RegistrationDatabase::clearAll): Deleted.
(WebCore::RegistrationDatabase::doPushChanges): Deleted.
(WebCore::RegistrationDatabase::doPushChangesWithOpenDatabase): Deleted.
(WebCore::RegistrationDatabase::importRecords): Deleted.
(WebCore::RegistrationDatabase::addRegistrationToStore): Deleted.
(WebCore::RegistrationDatabase::databaseFailedToOpen): Deleted.
(WebCore::RegistrationDatabase::databaseOpenedAndRecordsImported): Deleted.
* Source/WebCore/workers/service/server/RegistrationDatabase.h:
(WebCore::RegistrationDatabase::create): Deleted.
(WebCore::RegistrationDatabase::databaseDirectoryIsolatedCopy const): Deleted.
(): Deleted.
* Source/WebCore/workers/service/server/RegistrationStore.cpp:
(WebCore::RegistrationStore::RegistrationStore): Deleted.
(WebCore::RegistrationStore::~RegistrationStore): Deleted.
(WebCore::RegistrationStore::scheduleDatabasePushIfNecessary): Deleted.
(WebCore::RegistrationStore::pushChangesToDatabase): Deleted.
(WebCore::RegistrationStore::clearAll): Deleted.
(WebCore::RegistrationStore::flushChanges): Deleted.
(WebCore::RegistrationStore::startSuspension): Deleted.
(WebCore::RegistrationStore::closeDatabase): Deleted.
(WebCore::RegistrationStore::endSuspension): Deleted.
(WebCore::RegistrationStore::updateRegistration): Deleted.
(WebCore::RegistrationStore::removeRegistration): Deleted.
(WebCore::RegistrationStore::addRegistrationFromDatabase): Deleted.
(WebCore::RegistrationStore::didSaveWorkerScriptsToDisk): Deleted.
(WebCore::RegistrationStore::databaseFailedToOpen): Deleted.
(WebCore::RegistrationStore::databaseOpenedAndRecordsImported): Deleted.
* Source/WebCore/workers/service/server/RegistrationStore.h:
(WebCore::RegistrationStore::server): Deleted.
(WebCore::RegistrationStore::pushChangesToDatabase): Deleted.
(): Deleted.
* Source/WebCore/workers/service/server/SWRegistrationDatabase.cpp: Added.
(WebCore::databaseFilePath):
(WebCore::scriptDirectoryPath):
(WebCore::convertUpdateViaCacheToString):
(WebCore::convertStringToUpdateViaCache):
(WebCore::convertWorkerTypeToString):
(WebCore::convertStringToWorkerType):
(WebCore::currentRecordsTableSchema):
(WebCore::currentRecordsTableSchemaAlternate):
(WebCore::stripScriptSources):
(WebCore::populateScriptSourcesFromDisk):
(WebCore::SWRegistrationDatabase::statementString const):
(WebCore::SWRegistrationDatabase::cachedStatement):
(WebCore::SWRegistrationDatabase::SWRegistrationDatabase):
(WebCore::SWRegistrationDatabase::~SWRegistrationDatabase):
(WebCore::SWRegistrationDatabase::scriptStorage):
(WebCore::SWRegistrationDatabase::prepareDatabase):
(WebCore::SWRegistrationDatabase::ensureValidRecordsTable):
(WebCore::SWRegistrationDatabase::importRegistrations):
(WebCore::SWRegistrationDatabase::updateRegistrations):
* Source/WebCore/workers/service/server/SWRegistrationDatabase.h: Added.
* Source/WebCore/workers/service/server/SWRegistrationStore.h: Added.
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::storeRegistrationsOnDisk):
(WebCore::SWServer::SWServer):
(WebCore::SWServer::Connection::storeRegistrationsOnDisk):
(WebCore::SWServer::startSuspension): Deleted.
(WebCore::SWServer::endSuspension): Deleted.
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerDelegate.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::prepareToSuspend):
(WebKit::NetworkProcess::processDidResume):
(WebKit::NetworkProcess::storeServiceWorkerRegistrations):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
(WebKit::NetworkSession::createUniqueRegistrationStore):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWRegistrationStore.cpp: Added.
(WebKit::WebSWRegistrationStore::WebSWRegistrationStore):
(WebKit::WebSWRegistrationStore::clearAll):
(WebKit::WebSWRegistrationStore::flushChanges):
(WebKit::WebSWRegistrationStore::closeFiles):
(WebKit::WebSWRegistrationStore::importRegistrations):
(WebKit::WebSWRegistrationStore::updateRegistration):
(WebKit::WebSWRegistrationStore::removeRegistration):
(WebKit::WebSWRegistrationStore::scheduleUpdateIfNecessary):
(WebKit::WebSWRegistrationStore::updateToStorage):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWRegistrationStore.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::allManagedTypes):
(WebKit::NetworkStorageManager::closeServiceWorkerRegistrationFiles):
(WebKit::NetworkStorageManager::clearServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::importServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::updateServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::migrateServiceWorkerRegistrationsToOrigins):
(WebKit::NetworkStorageManager::updateServiceWorkerRegistrationsByOrigin):
(WebKit::NetworkStorageManager::shouldManageServiceWorkerRegistrationsByOrigin):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::toStorageType const):
(WebKit::OriginStorageManager::StorageBucket::toStorageIdentifier const):
(WebKit::OriginStorageManager::StorageBucket::serviceWorkerStorageManager):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::serviceWorkerStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/NetworkProcess/storage/ServiceWorkerStorageManager.cpp: Added.
(WebKit::ServiceWorkerStorageManager::ServiceWorkerStorageManager):
(WebKit::ServiceWorkerStorageManager::ensureDatabase):
(WebKit::ServiceWorkerStorageManager::closeFiles):
(WebKit::ServiceWorkerStorageManager::clearAllRegistrations):
(WebKit::ServiceWorkerStorageManager::importRegistrations):
(WebKit::ServiceWorkerStorageManager::updateRegistrations):
* Source/WebKit/NetworkProcess/storage/ServiceWorkerStorageManager.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _storeServiceWorkerRegistrations:]):
(-[WKWebsiteDataStore _hasRegisteredServiceWorker]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::storeServiceWorkerRegistrations):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Badging.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(log):
(install):
(addEventListener):
(then):
(catch):

Canonical link: https://commits.webkit.org/262631@main

webkit-commit-queue commented:
Committed 262631@main (1e35cedbb5b8): <https://commits.webkit.org/262631@main>

Reviewed commits have been landed. Closing PR #12204 and removing active labels.
